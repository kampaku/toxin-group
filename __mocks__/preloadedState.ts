const currentRooms = [
  {
    accessibility: { wideCorridor: true, assistant: false },
    additionalFacilities: {
      breakfast: true,
      crib: false,
      feedingChair: false,
      shampoo: false,
      tv: false,
      writingDesk: false,
    },
    freeDays: {
      from: '2022-10-11T06:17:02.467Z',
      to: '2022-10-18T00:20:15.895Z',
    },
    guests: { kids: 2, baby: 0, adults: 4 },
    id: 10,
    imgSrc: [
      'https://firebasestorage.googleapis.com/v0/b/toxin-…=media&token=43bbc6bb-a9a8-4731-ba0e-89784af2195e',
      'https://firebasestorage.googleapis.com/v0/b/toxin-…=media&token=43bbc6bb-a9a8-4731-ba0e-89784af2195e',
      'https://firebasestorage.googleapis.com/v0/b/toxin-…=media&token=43bbc6bb-a9a8-4731-ba0e-89784af2195e',
      'https://firebasestorage.googleapis.com/v0/b/toxin-…=media&token=43bbc6bb-a9a8-4731-ba0e-89784af2195e',
    ],
    number: 111,
    price: 11423,
    rating: 5,
    reviewsAmount: 99,
    roomAmenities: { bedrooms: 2, beds: 4, bathrooms: 0 },
    rules: { allowSmoke: false, allowGuests: true, allowPets: true },
    type: 'Люкс',
  },
  {
    accessibility: { wideCorridor: true, assistant: false },
    additionalFacilities: {
      breakfast: true,
      crib: false,
      feedingChair: false,
      shampoo: false,
      tv: false,
      writingDesk: false,
    },
    freeDays: {
      from: '2022-10-11T06:17:02.467Z',
      to: '2022-10-18T00:20:15.895Z',
    },
    guests: { kids: 2, baby: 0, adults: 4 },
    id: 20,
    imgSrc: [
      'https://firebasestorage.googleapis.com/v0/b/toxin-…=media&token=43bbc6bb-a9a8-4731-ba0e-89784af2195e',
      'https://firebasestorage.googleapis.com/v0/b/toxin-…=media&token=43bbc6bb-a9a8-4731-ba0e-89784af2195e',
      'https://firebasestorage.googleapis.com/v0/b/toxin-…=media&token=43bbc6bb-a9a8-4731-ba0e-89784af2195e',
      'https://firebasestorage.googleapis.com/v0/b/toxin-…=media&token=43bbc6bb-a9a8-4731-ba0e-89784af2195e',
    ],
    number: 222,
    price: 11423,
    rating: 5,
    reviewsAmount: 99,
    roomAmenities: { bedrooms: 2, beds: 4, bathrooms: 0 },
    rules: { allowSmoke: true, allowGuests: true, allowPets: false },
    type: 'Люкс',
  },
];

const filters = {
  accessibility: {
    wideCorridor: false,
    assistant: false,
  },
  additionalFacilities: {
    breakfast: false,
    crib: false,
    feedingChair: false,
    shampoo: false,
    tv: false,
    writingDesk: false,
  },
  freeDays: {
    from: null,
    to: null,
  },
  guests: { kids: 0, adults: 0, baby: 0 },
  price: { from: 5000, to: 10000 },
  roomAmenities: { bedrooms: 0, beds: 0, bathrooms: 0 },
  rules: { allowGuests: false, allowSmoke: false, allowPets: false },
};

const user = {
  id: '',
  name: '',
  surname: '',
  email: '',
  birthday: '',
  gender: '',
  getSpecOffers: false,
  likes: [],
};
  
const booking = {
  data: [
    {
      date: { from: '21.08.2022', to: '31.08.2022' },
      guests: { adults: 2, baby: 2, kids: 2 },
      totalCost: 2000,
      imgSrc: [''],
      number: 123,
      price: 20000,
      rating: 3,
      reviewsAmount: 10,
      type: '',
    },
  ],
  status: 'fulfilled',
};

export const preloadedState = {
  rooms: { current: currentRooms },
  searchFilters: filters,
  user,
  booking
};
