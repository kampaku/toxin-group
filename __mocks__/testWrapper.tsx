import React, { PropsWithChildren } from 'react';
import { configureStore, PreloadedState } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import { NextRouter } from 'next/router';
import { RouterContext } from 'next/dist/shared/lib/router-context';
import { render, RenderOptions } from '@testing-library/react';

import { createMockRouter } from './createMockRouter';

import { combinedReducer, RootState } from '../src/redux/store';

const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    reducer: combinedReducer,
    preloadedState,
  });
};

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: ReturnType<typeof setupStore>;
}

const renderWithProviders = (
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {},
  router: Partial<NextRouter> = {}
) => {
  const Wrapper = ({ children }: PropsWithChildren<{}>): JSX.Element => {
    return (
      <Provider store={store}>
        <RouterContext.Provider value={createMockRouter(router)}>
          {children}
        </RouterContext.Provider>
      </Provider>
    );
  };
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};

export { renderWithProviders };
