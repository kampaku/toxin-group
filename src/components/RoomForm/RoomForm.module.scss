@import 'scss/utils/vars.scss';
@import 'scss/utils/font-template.scss';

.roomForm {
  max-width: 380px;
  width: 100%;
  height: 512px;
  background: $colorBackground;
  border: 1px solid rgba(0, 0, 0, 0.12);
  box-shadow: 0 0 25px rgba(0, 0, 0, 0.2);
  border-radius: 4px;

  @media (max-width: $widthPhone) {
    height: fit-content;
  }

  &__inner {
    margin: 39px 30px 30px 30px;
    display: flex;
    flex-direction: column;
    gap: 20px;

    @media (max-width: $widthPhone) {
      width: 90%;
      margin: 0 auto;
      padding: 30px 5%;
    }
  }

  &__head {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: baseline;

    &Info {
      font-weight: bold;
    }

    &Sign {
      font-size: 14px;
      padding-right: 3px;
    }

    &Number {
      font-size: 24px;
    }

    &Type {
      @extend %label-text;
      color: $colorBrandMain;
      margin-left: 8px;
    }

    &Price {
      @extend %body-text;
      color: $dark-shade-50;
    }

    &Cost {
      line-height: 23px;
      margin-left: 4px;
    }

    &Period {
      font-size: $fontSizeLabel;
    }
  }

  &__guests {
    &Title {
      @extend %label-text;
      text-transform: uppercase;
      margin: 0px;
      padding-bottom: 6px;
    }
  }

  &__alert {
    @extend %body-text;
    padding: 30px;
  }
}
