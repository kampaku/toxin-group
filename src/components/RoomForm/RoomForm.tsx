import { FC, FormEvent, useCallback, useState } from 'react';
import Popup from 'reactjs-popup';
import { useTranslation } from 'next-i18next';

import { Button, Calendar, ConfigurationDropdown } from 'components';
import { useAppDispatch, useAuth, useBookingStatus } from 'hooks/hooks';
import { addBookingInfo } from 'redux/thunks/booking';
import { getPeriodBetweenDates } from 'utils/days';
import { setBookingStatus } from 'redux/slices/bookingSlice';

import { Calculation } from './Calculation/Calculation';
import styles from './RoomForm.module.scss';

type Props = {
  data: RoomCardType;
  discount?: number;
  serviceCharge?: number;
  additionalFees?: number;
  initialDate: FreeDays;
  initialGuests: Guests;
};

const RoomForm: FC<Props> = ({
  data,
  discount = 2179,
  serviceCharge = 0,
  additionalFees = 300,
  initialDate,
  initialGuests,
}) => {
  const dispatch = useAppDispatch();
  const { type, price, number } = data;
  const [date, setDate] = useState(initialDate);
  const [guests, setGuests] = useState(initialGuests);
  const [totalCost, setTotalCost] = useState(0);
  const hasDate = date.from !== null || date.to !== null;
  const hasGuests = guests.adults !== 0;
  const [openPopup, setOpenPopup] = useState(false);
  const status = useBookingStatus();
  const { t } = useTranslation('room-form');

  const userId = useAuth().id;
  const handleSelectedDays = useCallback((selectedDates: Date[]) => {
    const selectedDays = {
      from: selectedDates[0] ? selectedDates[0].toISOString() : null,
      to: selectedDates[1] ? selectedDates[1].toISOString() : null,
    };
    setDate(selectedDays);
  }, []);
  const handleGuestsChange = (content: Content) => {
    setGuests(content as Guests);
  };
  const handleSetCost = (cost: number) => setTotalCost(cost);
  const handleClosePopup = () => {
    setOpenPopup(false);
    dispatch(setBookingStatus('idle'));
  };

  const handleSubmit = (e: FormEvent) => {
    setOpenPopup(true);
    e.preventDefault();
    setOpenPopup(true);
    if (userId && hasDate && hasGuests) {
      dispatch(
        addBookingInfo({
          id: userId,
          data: { ...data, guests, date, totalCost },
        })
      );
    }
  };

  return (
    <form className={styles.roomForm} onSubmit={handleSubmit}>
      <div className={styles.roomForm__inner}>
        <div className={styles.roomForm__head}>
          <div className={styles.roomForm__headInfo}>
            <span className={styles.roomForm__headSign}>№</span>
            <span className={styles.roomForm__headNumber}>{number}</span>
            {type && (
              <span className={styles.roomForm__headType}>{t('luxury')}</span>
            )}
          </div>
          <div className={styles.roomForm__headPrice}>
            <strong className={styles.roomForm__headCost}>
              {price.toLocaleString()}₽ {t('perDay')}
            </strong>
          </div>
        </div>
        <div className={styles.roomForm__date}>
          <Calendar
            mode="twin"
            onSelectedDate={handleSelectedDays}
            selectedDates={initialDate}
          />
        </div>
        <div className={styles.roomForm__guests}>
          <ConfigurationDropdown
            title="guests"
            onChange={handleGuestsChange}
            initialContent={initialGuests}
          />
        </div>
        <Calculation
          price={price}
          discount={discount}
          serviceCharge={serviceCharge}
          additionalFees={additionalFees}
          days={getPeriodBetweenDates(date)}
          countCost={handleSetCost}
        />
        <div className={styles.roomForm__button}>
          <Button type="submit">{t('bookBtn')}</Button>
        </div>
        <Popup
          open={status === 'fulfilled'}
          closeOnDocumentClick
          onClose={handleClosePopup}
        >
          <span className={styles.roomForm__alert}>{t('successBook')}</span>
        </Popup>
        <Popup
          open={userId.length === 0 && openPopup}
          closeOnDocumentClick
          onClose={handleClosePopup}
        >
          <span className={styles.roomForm__alert}>{t('rejectBook')}</span>
        </Popup>
        <Popup
          open={userId.length > 0 && openPopup && !hasDate}
          closeOnDocumentClick
          onClose={handleClosePopup}
        >
          <span className={styles.roomForm__alert}>Выберите дату</span>
        </Popup>
        <Popup
          open={userId.length > 0 && openPopup && !hasGuests}
          closeOnDocumentClick
          onClose={handleClosePopup}
        >
          <span className={styles.roomForm__alert}>
            Выберите количество гостей
          </span>
        </Popup>
      </div>
    </form>
  );
};

export { RoomForm };
