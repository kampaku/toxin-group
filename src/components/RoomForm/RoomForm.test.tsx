import { screen } from '@testing-library/react';

import { renderWithProviders } from '../../../__mocks__/testWrapper';
import { RoomForm } from './RoomForm';

describe('Sign-in page', () => {
  let form: HTMLElement;

  const expectedProps = {
    data: {
      id: 100,
      imgSrc: [''],
      number: 130,
      type: '',
      rating: 3,
      price: 4000,
      costRange: '',
      reviewsAmount: 50,
    },
    initialDate: {
      from: new Date(2022, 6, 6).toISOString(),
      to: new Date(2022, 6, 8).toISOString(),
    },
    initialGuests: { adults: 1, baby: 0, kids: 0 },
  };

  beforeEach(() => {
    form = renderWithProviders(
      <RoomForm {...expectedProps} />,
      { preloadedState: {} },
      { asPath: '/' }
    ).container;
  });

  test('Sign in snapshot', () => {
    expect(form).toMatchSnapshot();
  });

  test('Room number exist', () => {
    const roomNumber = screen.getByText(/130/i);
    expect(roomNumber).toBeInTheDocument();
  });

  test('Calculation exist', () => {
    const calculation = screen.getByText(/2 day/i);
    expect(calculation).toBeInTheDocument();
  });
});
