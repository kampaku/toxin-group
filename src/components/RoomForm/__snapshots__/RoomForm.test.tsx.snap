// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign-in page Sign in snapshot 1`] = `
<div>
  <form
    class="roomForm"
  >
    <div
      class="roomForm__inner"
    >
      <div
        class="roomForm__head"
      >
        <div
          class="roomForm__headInfo"
        >
          <span
            class="roomForm__headSign"
          >
            №
          </span>
          <span
            class="roomForm__headNumber"
          >
            130
          </span>
        </div>
        <div
          class="roomForm__headPrice"
        >
          <strong
            class="roomForm__headCost"
          >
            4 000
            ₽ 
            perDay
          </strong>
        </div>
      </div>
      <div
        class="roomForm__date"
      >
        <div
          class="calendar"
        >
          <div
            class="wrapper"
          >
            <label
              class="label"
            >
              labelFirstTwin
              <input
                class="input"
                placeholder="placeholderTwin"
                readonly=""
                type="text"
              />
              <svg
                class="expandIcon"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                />
              </svg>
            </label>
            <label
              class="label"
            >
              labelSecond
              <input
                class="input"
                placeholder="placeholderTwin"
                readonly=""
                type="text"
              />
              <svg
                class="expandIcon"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                />
              </svg>
            </label>
          </div>
        </div>
      </div>
      <div
        class="roomForm__guests"
      >
        <div
          class="roomDropdown"
        >
          <p
            class="roomDropdown__title"
          >
            guests
          </p>
          <div
            class="roomDropdownInput"
            role="button"
            tabindex="0"
          >
            <div
              class=""
            >
              <input
                class="input input_isReadonly"
                name=""
                placeholder="Email"
                readonly=""
                type="text"
                value="1 guests"
              />
            </div>
            <span
              class="roomDropdownInput__iconWrapper"
            >
              <svg
                class="roomDropdownInput__icon"
                fill="none"
                height="20px"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="20px"
                xmlns="http://www.w3.org/2000/svg"
              >
                <polyline
                  points="6 9 12 15 18 9"
                />
              </svg>
            </span>
          </div>
        </div>
      </div>
      <div
        class="calculation"
      >
        <div
          class="calculation__inner"
        >
          <div
            class="calculation__row calculation__row_withoutInfo"
          >
            <div
              class="calculation__daysPriceText"
            >
              4 000
              ₽ х 
              2
               
              day
            </div>
            <div
              class="calculation__cost"
            >
              8 000
              ₽
            </div>
          </div>
          <div
            class="calculation__row"
          >
            <div
              class="calculation__serviceText"
            >
              charge
              : discount  2 179₽
            </div>
            <button
              aria-describedby="popup-1"
              class="calculation__info"
              type="button"
            >
              i
            </button>
            <div
              class="calculation__cost"
            >
              0
              ₽
            </div>
          </div>
          <div
            class="calculation__row"
          >
            <div
              class="calculation__additionalText"
            >
              additionalCharge
            </div>
            <button
              aria-describedby="popup-2"
              class="calculation__info"
              type="button"
            >
              i
            </button>
            <div
              class="calculation__cost"
            >
              300
              ₽
            </div>
          </div>
        </div>
        <div
          class="calculation__total"
        >
          <div
            class="calculation__totalText"
          >
            total
          </div>
          <div
            class="calculation__totalDots"
          >
             
          </div>
          <div
            class="calculation__totalCost"
          >
            6 121
            ₽
          </div>
        </div>
      </div>
      <div
        class="roomForm__button"
      >
        <button
          class="button button_type_submit"
          type="submit"
        >
          <span
            class="button__icon"
          >
            <svg
              fill="currentColor"
              height="24px"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="24px"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M12 4l-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
              />
            </svg>
          </span>
          <span
            class="button_type_signInText"
          >
            bookBtn
          </span>
        </button>
      </div>
    </div>
  </form>
</div>
`;
